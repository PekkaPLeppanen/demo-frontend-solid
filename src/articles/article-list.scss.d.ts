declare const styles: {
	readonly 'articleList': string;
};
export = styles;
