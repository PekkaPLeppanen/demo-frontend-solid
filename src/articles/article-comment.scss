.article-comment {
  border-top: 1px solid #dddddd;
  padding: 0.5em;

  &__email {
    color: #666;
    font-size: 0.75em;
  }

  &__body {
    padding: 0 0.5em;
    font-size: 0.875em;
  }
}
