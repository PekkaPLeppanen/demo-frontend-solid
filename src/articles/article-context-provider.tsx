import { createContext, createSignal, type FlowProps, type JSX, type Signal, useContext } from 'solid-js'
import { type IArticle } from '@/resources/article-provider'

type IArticleContext = Signal<IArticle | undefined>

/**
 * What is returned here can be accessed in the child components of the ArticleProvider through useArticle()
 */
const makeArticleContext = (article?: IArticle): IArticleContext => {
  return createSignal(article)
}

const ArticleContext = createContext<IArticleContext>(makeArticleContext())

/**
 * Creates a context provider component, where all the children have access to the ArticleContext
 *
 * @see https://www.solidjs.com/docs/latest/api#createcontext
 *
 * @param props
 * @constructor
 */
export function ArticleContextProvider (props: FlowProps): JSX.Element {
  return (
    <ArticleContext.Provider value={makeArticleContext()}>
      {props.children}
    </ArticleContext.Provider>
  )
}

/**
 * Usage: const [isMenuOpen, {openMenu, closeMenu}] = useArticle();
 */
export function useArticle (): IArticleContext {
  return useContext(ArticleContext)
}
