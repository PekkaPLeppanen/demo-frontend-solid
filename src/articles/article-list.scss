.article-list {
  list-style: none;

  &__item {
    text-transform: capitalize;
    padding: 1em;
    margin-top: 1em;
    border: 1px solid #ddd;
    border-radius: 5px;
  }
}
