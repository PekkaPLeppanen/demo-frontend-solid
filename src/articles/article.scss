.article {
  padding: 0 0.5em;

  &__title {}

  &__body {
    &:after {
      content: '.'
    }
  }
}
