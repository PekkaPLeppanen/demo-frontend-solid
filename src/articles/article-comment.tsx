import { type Component } from 'solid-js'
import { type IArticleComment } from '@/resources/article-provider'
import '@/articles/article-comment.scss'

export interface ArticleCommentProps {
  comment: IArticleComment
}

export const ArticleComment: Component<ArticleCommentProps> = (props) => {
  return (
    <div class="article-comment">
      <div class="article-comment__email">{props.comment.email}</div>
      <div class="article-comment__body">{props.comment.body}</div>
    </div>
  )
}
