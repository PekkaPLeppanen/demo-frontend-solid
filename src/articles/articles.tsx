import {Component, createResource, For} from 'solid-js';
import {Article, getAllArticles} from '@/resources/article-provider';
import {Suspense} from 'solid-js/web';
import '@/articles/article-list.scss';

export const Articles: Component = () => {

	const [articles] = createResource(getAllArticles);

	return (
		<>
			<h1>Latest topics</h1>
			<Suspense fallback={<p>Loading...</p>}>
				<ul class="article-list">
					<For each={articles()}>{(article: Article) =>
						<li class="article-list__item">
							{article.title}
						</li>
					}</For>
				</ul>
			</Suspense>
		</>

	);
};
