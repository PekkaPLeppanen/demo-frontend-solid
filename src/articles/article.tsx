import { type Component, createResource, For } from 'solid-js'
import { getArticleComments, type IArticle, type IArticleComment } from '@/resources/article-provider'
import { Suspense } from 'solid-js/web'
import { ArticleComment } from '@/articles/article-comment'
import '@/articles/article.scss'
import { useArticle } from '@/articles/article-context-provider'

export interface ArticleProps {
  article: IArticle
}

export const Article: Component = () => {
  const [article] = useArticle()
  const [comments] = createResource(article()?.id, getArticleComments)

  return (
    <div class="article">
      <h2 class="article__title">{article()?.title}</h2>
      <p class="article__body">{article()?.body}</p>
      <Suspense fallback={<p>Loading comments...</p>}>
        <For each={comments()}>{(comment: IArticleComment) =>
          <ArticleComment comment={comment}/>
        }</For>
      </Suspense>
    </div>
  )
}
