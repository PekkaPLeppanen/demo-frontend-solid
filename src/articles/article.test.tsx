import {render} from '@solidjs/testing-library';
import {afterEach, beforeEach, describe, expect, test} from 'vitest';
import {articleCommentsResponseMock, articleResponseMock} from '@/resources/__mocks__/articles';
import {Article} from '@/articles/article';
import {getArticleComments} from '@/resources/article-provider';
import {ArticleComment, ArticleCommentProps} from '@/articles/article-comment';

vi.mock('@/resources/article-provider');
vi.mock('@/articles/article-comment');

const getArticleCommentsMock = vi.mocked(getArticleComments);

const ArticleCommentMock = vi.mocked(ArticleComment);
const articleCommentPropsSpy = vi.fn();
ArticleCommentMock.mockImplementation((props: ArticleCommentProps) => {
	articleCommentPropsSpy(props);
	return <div data-testid="article-comment"/>;
});
describe('<Article />', () => {

	beforeEach(() => {
		vi.useFakeTimers();
	});

	afterEach(() => {
		vi.useRealTimers();
		vi.clearAllMocks();
	});

	test('render', () => {
		getArticleCommentsMock.mockResolvedValueOnce([]);

		const article = articleResponseMock[0];
		const {container, unmount, queryByText} = render(() => <Article article={article}/>);
		expect(container.outerHTML).toMatchSnapshot();
		expect(container.querySelector('.article__title')?.textContent).toEqual(article.title);
		expect(container.querySelector('.article__body')?.textContent).toEqual(article.body);
		expect(queryByText('Loading comments...')).toBeInTheDocument();
		unmount();
	});
	test('loading comments', async () => {
		getArticleCommentsMock.mockResolvedValueOnce(articleCommentsResponseMock);

		const article = articleResponseMock[0];
		const {queryByText, getAllByTestId, unmount} = render(() => <Article article={article}/>);
		await vi.advanceTimersToNextTimerAsync();
		expect(queryByText('Loading comments...')).not.toBeInTheDocument();
		expect(getAllByTestId('article-comment').length).toEqual(articleCommentsResponseMock.length);
		unmount();
	});
});
