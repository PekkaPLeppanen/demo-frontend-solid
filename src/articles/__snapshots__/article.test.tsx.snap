// Vitest Snapshot v1

exports[`<Article /> > render 1`] = `"<div><div class=\\"article\\"><h2 class=\\"article__title\\">sunt aut facere repellat provident occaecati excepturi optio reprehenderit</h2><p class=\\"article__body\\">quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto</p><p>Loading comments...</p></div></div>"`;
